adding the ability to delete a secret from LP

Yhoorneman opened the pull request.

Yhoorneman committed:
adding the ability to delete a secret from LP

costastf commented on lastpasslib/secrets.py:
```diff
@@ -186,6 +186,23 @@ def add_attachment(self, attachment):
+        response = self._lastpass.session.post(url, data=data)
+        if not response.ok:
+            response.raise_for_status()
+        self._lastpass.decrypted_vault.secrets = [secret for secret in self._lastpass.decrypted_vault.secrets if secret.name != self.name]
```
There can be duplicate names on secrets. I suggest we filter on ID since that is always unique.

Yhoorneman committed:
Added typing-extensions to comply with:
  File "/Users/yhoorneman/Git/lastpasslib/.venv/lib/python3.10/site-packages/astroid/constraint.py", line 19, in <module>
    from typing_extensions import Self
ModuleNotFoundError: No module named 'typing_extensions'

Yhoorneman committed:
- fixing linting errors
- missing time import

Yhoorneman committed:
imports

Yhoorneman committed:
add function delete_secret at lastpasslib level

Yhoorneman committed:
missed a line in selective commit

costastf commented on lastpasslib/secrets.py:
```diff
@@ -186,6 +187,27 @@ def add_attachment(self, attachment):
+        }
+        response = self._lastpass.session.post(url, data=data)
+        if not response.ok:
+            response.raise_for_status()
```
The conditional is not required, this will only raise if there is a not ok response.

costastf commented on lastpasslib/secrets.py:
```diff
@@ -186,6 +187,27 @@ def add_attachment(self, attachment):
+            'sentms': f"{time.time_ns() // 1_000_000}",
+            'token': self._lastpass.token
+        }
+        response = self._lastpass.session.post(url, data=data)
```
We should add a time out to all requests calls.

costastf commented on lastpasslib/lastpasslib.py:
```diff
@@ -653,7 +654,19 @@ def get_passwords(self):
+
+        Returns:
+            bool: True on success, raises an HTTPError exception otherwise.
+
```
Let's expand on the documentation on what this raises so it looks nice on the docs.

costastf commented on lastpasslib/lastpasslib.py:
```diff
@@ -653,7 +654,19 @@ def get_passwords(self):
+            bool: True on success, raises an HTTPError exception otherwise.
+
+        """
+        secret = self.get_secret_by_name(name)
```
This will raise an exception if multiple secrets exist with the same name, this should be in the documentation to expose that functionality to the user.

costastf commented on lastpasslib/lastpasslib.py:
```diff
@@ -653,7 +654,19 @@ def get_passwords(self):
+        """Deletes a secret from the vault.
+
+        Returns:
+            bool: True on success, raises an HTTPError exception otherwise.
```
It does not raise an HTTP exception.

costastf commented on lastpasslib/lastpasslib.py:
```diff
@@ -653,7 +654,19 @@ def get_passwords(self):
 
         """
-        return self.get_secrets(filter_='Password')
+        return self.get_secrets(filter_="Password")
```
I think that the project uses single quotes consistently, let's not start diverging here for no benefit.

costastf commented:
Some comments added

Yhoorneman committed:
updates based on comments
+
merged commits from main

Yhoorneman committed:
update of secrets based on comments

costastf commented on lastpasslib/lastpasslib.py:
```diff
@@ -666,6 +667,25 @@ def get_passwords(self):
+        Raises:
+            MultipleInstances: If more than one password is found with the same name.
+            HTTPError: If the deletion of the secret failed. 
+
```
The code would also raise InvalidSecret.

costastf commented on lastpasslib/secrets.py:
```diff
@@ -198,8 +219,7 @@ def shared_to_people(self):
-            if not response.ok:
-                response.raise_for_status()
+            response.raise_for_status()
             sent = response.json().get('sent')
```
Would we want to raise here or just log an error, since we return a boolean? I lean more towards the not raising to be honest...

Yhoorneman committed:
delete secret by name now only returns a bool, as advised in comments

Yhoorneman committed:
adding delete_secret_by_id

Yhoorneman committed:
adding functions: get_password_by_name, get_password_by_id

Yhoorneman committed:
adding functions: delete_password_by_name, delete_secret_by_id

costastf commented on lastpasslib/lastpasslib.py:
```diff
@@ -658,6 +657,45 @@ def get_secrets_shared_directly(self):
+            self._logger.error(f'Secret with id "{id}" not found.')
+            return False
+        return secret.delete()
+    # TODO: implement delete secret_by_id, password, secure note, etc.
```
This is not needed any more i guess.

costastf reviewed (commented):
Maybe sort the methods by function, first the getting by and then the deleting?
Something like 
```
get_secret_by_name
get_secret_by_id
delete_secret_by_name
delete_secret_by_id
get_password_by_name
get_password_by_id
delete_password_by_name
delete_password_by_id
```

Yhoorneman committed:
Adding the functions for secure_note

Yhoorneman committed:
fixes after testing and linting

Yhoorneman committed:
putting filter_ back, since the argument is used in functions like: get_passwords_by_shared_folder
